[inductor] tensor_is_align fallbacking False if unbacked expr not comptime evaled

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133002
* __->__ #132423

Currently if storage_offset is unbacked symbol and is_align can not be computed compiletime - it hard fails.

Doing the best we can: adding guard_size_oblivious and fallback on False if can not be evaluated compiletime


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

IvanKobzarev committed:
[inductor] tensor_is_align fallbacking False if unbacked expr not comptime evaled

[ghstack-poisoned]

Microve commented on torch/_inductor/utils.py:
```diff
@@ -1812,9 +1812,15 @@ def tensor_is_aligned(tensor: torch.Tensor):
+    from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+
+    try:
+        return guard_size_oblivious(
```
Wonder whether we need `guard_size_oblivious` here. From my understanding, it can not prevent the exception because we don't know whether `storage_offset()` returns a even number of not

ezyang replied:
Yeah, size oblivious is wrong here. It probably doesn't matter, because all size oblivious will do is make us assume storage_offset isn't zero or one, which won't affect whether or not we can determine if the modulus is zero. But better to drop it.

ezyang reviewed (changes requested):
Need a test. I am suspicious this change is actually doing anything.

ezyang commented on torch/_inductor/utils.py:
```diff
@@ -1812,9 +1812,15 @@ def tensor_is_aligned(tensor: torch.Tensor):
+            == 0
+        )
+    except torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode:
+        return False
```
Don't catch the exception. Instead, more directly do something like statically_known_true

Microve replied:
Just want to make sure my understanding is correct:  For this case, since we can not evaluate the expr in a static way, so statically_known_true is just to return False, but during runtime, statically_known_true will use the real values to evaluate. 

If so, let's assume that during runtime, the value returned by statically_known_true is True, what will happen? Since 
during compilation,  statically_known_true is False, so the compiled graph is based on "statically_known_true is False", but during runtime, statically_known_true becomes True, then do we need to recompile?

Microve replied:
@ezyang any idea?

ezyang replied:
I mean, this is a context dependent question. When you are testing this to determine if you can apply an optional optimization, such a difference would be fine.

ezyang replied:
FWIW, your code here is also doing the same thing

Microve replied:
> I mean, this is a context dependent question. When you are testing this to determine if you can apply an optional optimization, such a difference would be fine.

A bit confused here, do you still need to recompile or not for the case I described?

Microve replied:
> FWIW, your code here is also doing the same thing

Which code? Could you elaborate a bit more?

ezyang replied:
You don't necessarily have to recompile.

Let me give an example. Suppose you are compiling 

```
def f(x: int, y: int):
    return x + (y % 2)
```

You are considering an optimization to optimize this into:

```
def f(x: int, y: int):
    return x
```

Clearly, this optimization is only valid when y % 2 == 0. So you want to test if this expression is true. If it is statically known to be true, the optimization is safe. But suppose we do not know. It is safe not to optimize the program and keep the original, because the original program is valid for both cases.

Microve replied:
I see. Just want to make sure I understand correct:
```
def f(x: int, y: int):
    if y % 2 == 0:
       return x + 1
    else:
       return x
```
Assume `y` is an unbacked symint, so dynamo throws an error. Then we put `statically_known_true` around the statement as below
```
def f(x: int, y: int):
    if statically_known_true(y % 2 == 0):
       return x + 1
    else:
       return x
```
if dynamo can figure out the value of the statement by static analysis, let's say y is even, then codes will be simplified as:
```
def f(x: int, y: int):
       return x + 1
```
if it can not figure out the value, then dynamo keep it as it is:

```
def f(x: int, y: int):
    if y % 2 == 0:
       return x + 1
    else:
       return x
```
Is that correct?

ezyang replied:
No, not really. You're not "put[ting] statically_known_true around the statement as below", this is happening entirely in the optimizer.

Microve replied:
> No, not really. You're not "put[ting] statically_known_true around the statement as below", this is happening entirely in the optimizer.

Hmm, did not get it. In this PR we use `statically_known_true` to resolve the issue, why in the above example, we don't use `statically_known_true`, is it because these are user codes not inside inductor?

Is `statically_known_true(x)` to give a hint to the compiler that `x` is True, or it asks the compiler to do static analysis on `x` to determine whether it is true or not?

ezyang replied:
You're mixing up user code (the code sample we were looking at) and Inductor lowering code (where the statically_known_true shows up)

Microve replied:
@ezyang I see, but what does `statically_known_true(x)` actually do?  Let's assume that if inductor wants to apply an optimization, it sees `statically_known_true(x)`. Does it tell inductor that "try your best to do static analysis about x, but if you are uncertain about x's value, then don't apply the optimization"?

ezyang replied:
you are saying things that individually are right but they are not being put together correctly. If you have a symbolic expression, you use statically_known_true in Inductor to do this analysis. Then you use the result to make a decision about whether or not to do an optimization or not

Microve replied:
@ezyang let's assume that during compile time `statically_known_true(x)` returns False, and compiler decides not to do optimization, but during run-time, `statically_known_true(x)` returns True, will compiler apply the optimization, If it will, then it seems like that the compile's decision at compile time does not matter which could be changed later during run-time, is that the case?

ezyang replied:
That's not how it works. The statically_known_true test is a compile time concept. At runtime we never call "statically_known_true".

Microve replied:
@ezyang I see, if we only do it during compile time, does that mean that if we can not reason the value of `x` during compile time, we will not do optimization. But `x` could be True during runtime, so it misses the optimization opportunity.

ezyang replied:
I mean, it's debatable whether or not there was an "optimization opportunity" at all in the first place. Is changing "if x == 2 then x * 2 else x + 5" to "x * 2" and optimization? Not really, it's just wrong LOL

IvanKobzarev committed:
Update on "[inductor] tensor_is_align fallbacking False if unbacked expr not comptime evaled"


Currently if storage_offset is unbacked symbol and is_align can not be computed compiletime - it hard fails.

Doing the best we can: adding guard_size_oblivious and fallback on False if can not be evaluated compiletime


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor] tensor_is_align fallbacking False if unbacked expr not comptime evaled"


Currently if storage_offset is unbacked symbol and is_align can not be computed compiletime - it hard fails.

Doing the best we can: adding guard_size_oblivious and fallback on False if can not be evaluated compiletime


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[inductor] tensor_is_align fallbacking False if unbacked expr not comptime evaled"


Currently if storage_offset is unbacked symbol and is_align can not be computed compiletime - it hard fails.

Doing the best we can: adding guard_size_oblivious and fallback on False if can not be evaluated compiletime


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

IvanKobzarev commented:
> Need a test. I am suspicious this change is actually doing anything.

@ezyang 
I've tried to make a test, but failed. How could I make input that has unbacked storage_offset?
What I tried:
```
    @torch._dynamo.config.patch(capture_scalar_outputs=True)
    @torch._inductor.config.patch(implicit_fallbacks=True)
    def test_non_comptime_known_input_is_aligned(self):
        with torch.library._scoped_library("mylib", "FRAGMENT") as lib:
            lib.define("foo(Tensor x) -> Tensor")

            def foo_impl(a):
                return a

            for backend in ["CPU", "CUDA", "Meta"]:
                lib.impl("foo", foo_impl, backend)                

            def fn(x):
                u = x.size(0)
                torch._check_is_size(u)
                torch._check(u > 0)
                torch._check(u < 8)
                t = inp.narrow(1, u, 8)
                return torch.ops.mylib.foo(t)

            
            base = torch.randn(4 * 64 + 64, dtype=torch.float32, device=GPU_TYPE)
            offset = 2
            inp = torch.as_strided(base, (4, 64), (64, 1), offset)

            ref_out = fn(inp)

            torch._dynamo.decorators.mark_unbacked(inp, 0)
            torch._dynamo.decorators.mark_static(inp, 1)

            out = torch.compile(fn, backend="inductor", fullgraph=True, dynamic=True)(inp)

            torch.testing.assert_allclose(ref_out, out)
```
Inductor checks is_aligned only for inputs, so I experimented with custom_op, but custom_op inputs do not have this requirement on alignment.

I guess we got it in full model compilation because of graph breaks.

ezyang commented:
You have to do a custom op. Check test_dynamic_stride_nobreak in test/inductor/test_torchinductor_dynamic_shapes.py for an example

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge
